src/vehicle_move.cpp

```diff
@@ -907,9 +932,65 @@ void vehicle::autodrive( int x, int y )
     }
 }
 
-void vehicle::pldrive( const point &p )
+bool vehicle::check_is_heli_landed()
+{
+    // @TODO - when there are chasms that extend below z-level 0 - perhaps the heli
+    // will be able to descend into them but for now, assume z-level-0 == the ground.
+    if( global_pos3().z == 0 || !g->m.has_flag_ter_or_furn( TFLAG_NO_FLOOR, global_pos3() ) ) {
+        is_flying = false;
+        return true;
+    }
+    return false;
+}
+
+bool vehicle::check_heli_descend( player &p )
+{
+    if( !is_rotorcraft() ) {
+        debugmsg( "A vehicle is somehow flying without being an aircraft" );
+        return true;
+    }
+    for( const tripoint &pt : get_points( true ) ) {
+        tripoint below( pt.xy(), pt.z - 1 );
+        if( g->m.has_zlevels() && ( pt.z < -OVERMAP_DEPTH ||
+                                    !g->m.has_flag_ter_or_furn( TFLAG_NO_FLOOR, pt ) ) ) {
+            p.add_msg_if_player( _( "You are already landed!" ) );
+            return false;
+        }
+        const optional_vpart_position ovp = g->m.veh_at( below );
+        if( g->m.impassable_ter_furn( below ) || ovp || g->critter_at( below ) ) {
+            p.add_msg_if_player( m_bad,
+                                 _( "It would be unsafe to try and land when there are obstacles below you." ) );
+            return false;
+        }
+    }
+    if( velocity > 0 ) {
+        p.add_msg_if_player( m_bad, _( "It would be unsafe to try and land while you are moving." ) );
+        return false;
+    }
+    return true;
+
+}
+
+bool vehicle::check_heli_ascend( player &p )
+{
+    if( !is_rotorcraft() ) {
+        debugmsg( "A vehicle is somehow flying without being an aircraft" );
+        return true;
+    }
+    if( velocity > 0 ) {
+        p.add_msg_if_player( m_bad, _( "It would be unsafe to try and take off while you are moving." ) );
```

Comment:
This prevents you from going to a different z level while flying normally, you must first stop, then you can go to the upper z level, then decelerate.